lib/blob-server/fastify-plugin.js

```diff
@@ -0,0 +1,120 @@
+import assert from 'node:assert'
+import fp from 'fastify-plugin'
+import { filetypemime } from 'magic-bytes.js'
+
+import { SUPPORTED_BLOB_VARIANTS } from '../blob-store/index.js'
+
+export default fp(blobServerPlugin, {
+  fastify: '4.x',
+  name: 'mapeo-blob-server',
+})
+
+/**
+ * @typedef {Object} BlobServerPluginOpts
+ *
+ * @property {import('../blob-store/index.js').BlobStore} blobStore
+ */
+
+/**
+ * @typedef {Object} ParsedParams
+ *
+ * @property {string} driveId
+ * @property {import('../types').BlobType} type
+ * @property {import('../types').BlobVariant<ParsedParams['type']>} variant
+ * @property {string} name
+ */
+
+const PARAMS_JSON_SCHEMA = /** @type {const} */ {
+  type: 'object',
+  required: ['driveId', 'type', 'variant', 'name'],
+  properties: {
+    driveId: { type: 'string' },
+    type: { type: 'string', enum: Object.keys(SUPPORTED_BLOB_VARIANTS) },
+    variant: {
+      type: 'string',
+      enum: [...new Set(Object.values(SUPPORTED_BLOB_VARIANTS).flat())],
+    },
+    name: { type: 'string' },
+  },
+}
+
+/** @type {import('fastify').FastifyPluginAsync<import('fastify').RegisterOptions & BlobServerPluginOpts>} */
+async function blobServerPlugin(fastify, options) {
+  if (!options.blobStore) throw new Error('Missing blobStore')
+
+  // We call register here so that the `prefix` option can work if desired
+  // https://fastify.dev/docs/latest/Reference/Routes#route-prefixing-and-fastify-plugin
+  fastify.register(routes, options)
+}
+
+/** @type {import('fastify').FastifyPluginAsync<Omit<BlobServerPluginOpts, 'prefix'>>} */
+async function routes(fastify, options) {
+  const { blobStore } = options
+
+  fastify.get(
+    '/:driveId/:type/:variant/:name',
+    { schema: { params: PARAMS_JSON_SCHEMA } },
+    /**
+     * @param {import('fastify').FastifyRequest<{ Params: ParsedParams }>} request
+     */
+    async (request, reply) => {
+      const { driveId, type, variant, name } = request.params
+
+      const blobId = constructBlobId(driveId, type, variant, name)
```

Comment:
wasn't keen on trying to figure out how to fully reflect the discriminate union of the BlobVariant in JSON Schema format, so using `contructBlobId` to assert and guarantee that the provided combination of `type` and `variant` are valid at runtime

Reply:
That's valid. I switched to a typescript type guard for slightly stricter typing.